CLOUDSTACK-10146: Bypass Secondary Storage for KVM templates

nvazquez opened the pull request:
JIRA Ticket: https://issues.apache.org/jira/browse/CLOUDSTACK-10146

### Background

This feature allows using templates and ISOs avoiding secondary storage as intermediate cache on KVM. The virtual machine deployment process is enhanced to supported bypassed registered templates and ISOs, delegating the work of downloading them to primary storage to the KVM agent instead of the SSVM agent.

### Description
#### Template and ISO registration: 
- When hypervisor is KVM, a checkbox is displayed with 'Direct Download' label.
![image](https://user-images.githubusercontent.com/5295080/34463109-5baa48e4-ee31-11e7-9eef-8e270ce00ddd.png)
- API methods `registerTemplate` and `registerISO` are both extended with this new parameter `directdownload`.
- On template or ISO registration, no download job is sent to SSVM agent, CloudStack would only persist an entry on `template_store_ref` indicating that template or ISO has been marked as 'Direct Download' (bypassing Secondary Storage). These entries are persisted as:
   - template_id = Template or ISO id on vm_template table
   - store_id NULL
   - download_state = BYPASSED
   - state = Ready
(Note: these entries allow users to deploy virtual machine from registered templates or ISOs)
- An URL validation command is sent to a random KVM host to check if template/ISO location can be reached. Metalink are also supported by this feature. In case of a metalink, it is fetched and URL check is performed on each of its URLs.
- Checksum should be provided as indicated on #2246: `{ALGORITHM}CHKSUMHASH`
- After template or ISO is registered, it would be displayed in the UI as:
![image](https://user-images.githubusercontent.com/5295080/34463248-de728ee2-ee33-11e7-90ad-042475cc96e8.png)

![image](https://user-images.githubusercontent.com/5295080/34463246-cc195226-ee33-11e7-90c9-3595cad6050a.png)

#### Virtual machine deployment
- When a 'Direct Download' template is selected for deployment, CloudStack would delegate template downloading to destination storage pool via destination host by a new pluggable download manager. 
- Download manager would handle template downloading depending on URL protocol. In case of HTTP, request headers can be set by the user via vm_template_details. Those details should be persisted as:
   - Key: HTTP_HEADER
   - Value: HEADERNAME:HEADERVALUE
In case of HTTPS, a new API method is added `uploadTemplateDirectDownloadCertificate` to allow user importing a client certificate into all KVM hosts' keystore before deployment.
- After template or ISO is downloaded to primary storage, usual entry would be persisted on `template_spool_ref` indicating the mapping between template/ISO and storage pool.

nvazquez commented:
Ping for review @rhtyd @DaanHoogland @rafaelweingartner

nvazquez commented:
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1530

nvazquez commented:
@blueorangutan test

blueorangutan commented:
@nvazquez a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-1959)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 33785 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2379-t1959-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 66 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1549

borisstoyanov commented:
@blueorangutan test matrix

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins matrix job (centos6 mgmt + xs65sp1, centos7 mgmt + vmware55u3, centos7 mgmt + kvmcentos7) has been kicked to run smoke tests

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1558

nvazquez commented:
@rhtyd @borisstoyanov @DaanHoogland marvin tests added, results on my local environment:

````
cat /tmp//MarvinLogs/test_templates_GEJ8QG/results.txt 
test_09_copy_delete_template (integration.smoke.test_templates.TestCopyDeleteTemplate) ... SKIP: Skipping test due to there are less than two zones.
Test create public & private template ... === TestName: test_01_create_template | Status : SUCCESS ===
ok
Test when createTemplate is used to create templates having the same name all of them get ... === TestName: test_CreateTemplateWithDuplicateName | Status : SUCCESS ===
ok
test_02_1_create_template_with_checksum_sha1_negative (integration.smoke.test_templates.TestCreateTemplateWithChecksum) ... === TestName: test_02_1_create_template_with_checksum_sha1_negative | Status : SUCCESS ===
ok
test_02_create_template_with_checksum_sha1 (integration.smoke.test_templates.TestCreateTemplateWithChecksum) ... === TestName: test_02_create_template_with_checksum_sha1 | Status : SUCCESS ===
ok
test_03_1_create_template_with_checksum_sha256_negative (integration.smoke.test_templates.TestCreateTemplateWithChecksum) ... === TestName: test_03_1_create_template_with_checksum_sha256_negative | Status : SUCCESS ===
ok
test_03_create_template_with_checksum_sha256 (integration.smoke.test_templates.TestCreateTemplateWithChecksum) ... === TestName: test_03_create_template_with_checksum_sha256 | Status : SUCCESS ===
ok
test_04_1_create_template_with_checksum_md5_negative (integration.smoke.test_templates.TestCreateTemplateWithChecksum) ... === TestName: test_04_1_create_template_with_checksum_md5_negative | Status : SUCCESS ===
ok
test_04_create_template_with_checksum_md5 (integration.smoke.test_templates.TestCreateTemplateWithChecksum) ... === TestName: test_04_create_template_with_checksum_md5 | Status : SUCCESS ===
ok
test_05_create_template_with_no_checksum (integration.smoke.test_templates.TestCreateTemplateWithChecksum) ... === TestName: test_05_create_template_with_no_checksum | Status : SUCCESS ===
ok
Register a template using Direct Download flag ... === TestName: test_01_register_template_direct_download_flag | Status : SUCCESS ===
ok
Deploy a VM from a Direct Download registered template ... === TestName: test_02_deploy_vm_from_direct_download_template | Status : SUCCESS ===
ok
Deploy a VM from a Direct Download registered template with wrong checksum ... === TestName: test_03_deploy_vm_wrong_checksum | Status : SUCCESS ===
ok
Test Edit template ... === TestName: test_02_edit_template | Status : SUCCESS ===
ok
Test delete template ... === TestName: test_03_delete_template | Status : SUCCESS ===
ok
Test for extract template ... === TestName: test_04_extract_template | Status : SUCCESS ===
ok
Update & Test for template permissions ... === TestName: test_05_template_permissions | Status : SUCCESS ===
ok
Test for copy template from one zone to another ... SKIP: Not enough zones available to perform copy template
Test only public templates are visible to normal user ... === TestName: test_07_list_public_templates | Status : SUCCESS ===
ok
Test System templates are not visible to normal user ... === TestName: test_08_list_system_templates | Status : SUCCESS ===
ok

----------------------------------------------------------------------
Ran 20 tests in 1373.580s

OK (SKIP=2)
````

blueorangutan commented:
<b>Trillian test result (tid-1980)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 30878 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2379-t1980-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_privategw_acl.py
Smoke tests completed. 65 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_04_rvpc_privategw_static_routes | `Failure` | 270.09 | test_privategw_acl.py

rafaelweingartner commented:
@nvazquez before I start reviewing I have some doubts. Can you answer them?
I read your PR description, which is nicely detailed (as always a very good work!). However, I have some more specific doubts. 

* Is it specific for KVM (I am assuming it is)? 

* Will the template/ISO be downloaded as soon as you register it? Or only when there is the need for such volume/ISO?

* After the template/ISO is downloaded, if there is no VMs instantiated from this template or no VMs using the ISO, is the volume(ISO/volume of the template) cleaned up?

* As soon as we delete a template in ACS, is the file deleted in that moment? Or is it only marked as “Destroyed”, and just later a thread will remove it?

nvazquez commented:
Thanks for your time @rafaelweingartner, let me answer your questions:
- Yes, this feature is specific for KVM
- Template/ISO does not get downloaded when registered. Download is performed when they are needed.
- No, once it gets downloaded it would be there unless explicitly removed
- It does not get deleted at that moment, it gets marked for garbage collection on template_spool_ref and then physically deleted when gc task executes

blueorangutan commented:
<b>Trillian test result (tid-1981)</b>
Environment: vmware-55u3 (x2), Advanced Networking with Mgmt server 7
Total time taken: 37355 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2379-t1981-vmware-55u3.zip
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Smoke tests completed. 66 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

blueorangutan commented:
<b>Trillian test result (tid-1979)</b>
Environment: xenserver-65sp1 (x2), Advanced Networking with Mgmt server 6
Total time taken: 43379 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2379-t1979-xenserver-65sp1.zip
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Smoke tests completed. 66 look OK, 0 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---

rhtyd commented:
Tests LGTM. I'll wait for lgtm from @borisstoyanov @DaanHoogland @rafaelweingartner

marcaurele commented on agent/src/com/cloud/agent/direct/download/HttpDirectTemplateDownloader.java:
```diff
@@ -0,0 +1,158 @@
+        RandomAccessFile out = null;
+        try {
+            client.executeMethod(request);
+            remoteSize = getRemoteSize(request);
```
`remoteSize` is never read

nvazquez replied:
Hi @marcaurele, in fact it is, on copyBytes method in line 141

marcaurele replied:
Since you're using this info in the `copyBytes` method, this is overkill. It can even throw an exception but having the filesize isn't mandatory to download it. This information, if present in the header, should be used as a *nice to have* but the download **must work** without this info as @wido pointed that some servers don't send this information in the header (https://github.com/apache/cloudstack/pull/1934#issuecomment-347784289). So it must be changed

marcaurele commented on agent/src/com/cloud/agent/direct/download/HttpDirectTemplateDownloader.java:
```diff
@@ -0,0 +1,158 @@
+        } finally {
+            try {
+                out.close();
+                in.close();
```
Would be nicer to move the code in a try with resource statement (https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html)

marcaurele commented on api/src/org/apache/cloudstack/api/command/user/iso/RegisterIsoCmd.java:
```diff
@@ -168,6 +173,10 @@ public Boolean isDynamicallyScalable() {
         return isDynamicallyScalable ==  null ? Boolean.FALSE : isDynamicallyScalable;
     }
 
+    public Boolean isDirectDownload() {
```
can return a `boolean` instead since it cannot be null

marcaurele commented on api/src/org/apache/cloudstack/api/command/user/template/RegisterTemplateCmd.java:
```diff
@@ -263,6 +269,10 @@ public Boolean isRoutingType() {
         return isRoutingType;
     }
 
+    public Boolean isDirectDownload() {
```
`boolean` here too

marcaurele commented on core/src/org/apache/cloudstack/agent/directdownload/HttpDirectDownloadCommand.java:
```diff
@@ -0,0 +1,46 @@
+
+    @Override
+    public void setExecuteInSequence(boolean inSeq) {
+    }
```
Function can be removed

marcaurele commented on core/src/org/apache/cloudstack/agent/directdownload/MetalinkDirectDownloadCommand.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public void setExecuteInSequence(boolean inSeq) {
+
+    }
```
Function can be removed

marcaurele commented on core/src/org/apache/cloudstack/agent/directdownload/NfsDirectDownloadCommand.java:
```diff
@@ -0,0 +1,38 @@
+    @Override
+    public void setExecuteInSequence(boolean inSeq) {
+
+    }
```
Function can be removed

marcaurele commented on engine/api/src/org/apache/cloudstack/engine/subsystem/api/storage/TemplateInfo.java:
```diff
@@ -25,4 +25,6 @@
 
     String getInstallPath();
+
+    boolean isDirectDownload();
```
Just for your info, be aware of not mixing `Boolean` and `boolean` types since the JVM won't check for those kind of mismatch as it does auto casting to the expected type. But you will end up with a NPE when getting a `Boolean` NULL on a method expecting a `boolean`.

nvazquez replied:
Thanks

marcaurele commented on engine/orchestration/src/org/apache/cloudstack/engine/orchestration/VolumeOrchestrator.java:
```diff
@@ -1273,6 +1273,7 @@ public void prepareForMigration(VirtualMachineProfile vm, DeployDestination dest
         for (int i = 0; i < 2; i++) {
             // retry one more time in case of template reload is required for Vmware case
             AsyncCallFuture<VolumeApiResult> future = null;
+            boolean bypassed = false;
```
unused variable

marcaurele commented on engine/schema/resources/META-INF/db/schema-41000to41100.sql:
```diff
@@ -516,3 +516,109 @@ UPDATE `cloud`.`vm_template` SET guest_os_id=99 WHERE id=8;
+ADD COLUMN `direct_download` TINYINT(1) DEFAULT '0' COMMENT 'Indicates if Secondary Storage is bypassed and template is downloaded to Primary Storage';
+
+DROP VIEW IF EXISTS `cloud`.`template_view`;
+CREATE VIEW `template_view` AS
```
Better to put those drop & create in a single statement:
```sql
CREATE OR REPLACE VIEW `template_view` AS
```

marcaurele commented on engine/schema/src/com/cloud/storage/VMTemplateVO.java:
```diff
@@ -605,6 +609,10 @@ public void setUpdated(Date updated) {
         this.updated = updated;
     }
 
+    public Boolean isDirectDownload() {
```
`boolean`

marcaurele commented on engine/storage/image/src/org/apache/cloudstack/storage/image/TemplateDataFactoryImpl.java:
```diff
@@ -153,4 +164,67 @@ public TemplateInfo getReadyTemplateOnCache(long templateId) {
+            List<VMTemplateStoragePoolVO> existingRefs = templatePoolDao.listByTemplateId(templateId);
+            pool = getOneMatchingPoolIdFromRefs(existingRefs, pools);
+        }
+        VMTemplateStoragePoolVO spoolRef = templatePoolDao.findByPoolTemplate(pool, templateId);
```
`pool` can be null here, according to https://github.com/apache/cloudstack/pull/2379/files#diff-92e0bb0fd44a51c2c2234ae65d88fbbdR181

marcaurele commented on engine/storage/volume/src/org/apache/cloudstack/storage/datastore/manager/PrimaryDataStoreProviderManagerImpl.java:
```diff
@@ -66,6 +67,11 @@ public PrimaryDataStore getPrimaryDataStore(long dataStoreId) {
 
+    private List<PrimaryDataStore> listPrimaryDataStoresByClusterId(Long clusterId) {
+        List<StoragePoolVO> storagePoolVOS = dataStoreDao.listPoolsByCluster(clusterId);
+        return null;
```
`return null` ? I guess it should be `return storagePoolVOS`

marcaurele commented on plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:
```diff
@@ -1387,4 +1405,35 @@ public Answer forgetObject(final ForgetObjectCmd cmd) {
+    public Answer handleDownloadTemplateToPrimaryStorage(DirectDownloadCommand cmd) {
+        final PrimaryDataStoreTO pool = cmd.getDestPool();
+        if (!pool.getPoolType().equals(StoragePoolType.NetworkFilesystem)) {
+            return new DirectDownloadAnswer(false, "Unsopported pool type " + pool.getPoolType().toString());
```
typo: `Unsupported pool type...`

marcaurele commented on plugins/hypervisors/kvm/src/com/cloud/hypervisor/kvm/storage/KVMStorageProcessor.java:
```diff
@@ -1387,4 +1405,35 @@ public Answer forgetObject(final ForgetObjectCmd cmd) {
+        } else if (cmd instanceof MetalinkDirectDownloadCommand) {
+            downloader = new MetalinkDirectTemplateDownloader(cmd.getUrl(), destPool.getLocalPath(), cmd.getTemplateId(), cmd.getChecksum());
+        } else {
+            return new DirectDownloadAnswer(false, "Unsopported protocol, please provide HTTP(S), NFS or a metalink");
```
Same typo here: `Unsupported...`

marcaurele commented on engine/schema/src/com/cloud/storage/VMTemplateVO.java:
```diff
@@ -146,6 +146,9 @@
     protected boolean dynamicallyScalable;
 
+    @Column(name = "direct_download")
+    private Boolean directDownload;
```
Since it's `0` by default and cannot be `null`, it can be switched to a `boolean`

marcaurele commented on server/src/com/cloud/api/query/vo/TemplateJoinVO.java:
```diff
@@ -222,6 +222,9 @@
     private String tempZonePair; // represent a distinct (templateId, data_center_id) pair
 
+    @Column(name = "direct_download")
+    private Boolean directDownload;
```
`boolean` here too, same comment as on the other VO

marcaurele commented on server/src/com/cloud/api/query/vo/TemplateJoinVO.java:
```diff
@@ -477,4 +480,7 @@ public void setAccountId(long accountId) {
         this.accountId = accountId;
     }
 
+    public Boolean getDirectDownload() {
```
`boolean`

marcaurele commented on server/src/com/cloud/template/TemplateAdapterBase.java:
```diff
@@ -339,9 +341,13 @@ protected VMTemplateVO persistTemplate(TemplateProfile profile, VirtualMachineTe
                 profile.getTemplateType(), profile.getUrl(), profile.getRequiresHVM(), profile.getBits(), profile.getAccountId(), profile.getCheckSum(),
                 profile.getDisplayText(), profile.getPasswordEnabled(), profile.getGuestOsId(), profile.getBootable(), profile.getHypervisorType(),
-                profile.getTemplateTag(), profile.getDetails(), profile.getSshKeyEnabled(), profile.IsDynamicallyScalable());
+                profile.getTemplateTag(), profile.getDetails(), profile.getSshKeyEnabled(), profile.IsDynamicallyScalable(), profile.isDirectDownload());
```
`profile.isDirectDownload()` might return null since all constructors don't set the direct download `Boolean` variable. The method declaration expects a `boolean`, so no `null`.

nvazquez replied:
Done, thanks

marcaurele commented on server/src/org/apache/cloudstack/direct/download/DirectDownloadManagerImpl.java:
```diff
@@ -0,0 +1,225 @@
+        SetupDirectDownloadCertificate cmd = new SetupDirectDownloadCertificate(certificate, certificateName);
+        Answer answer = agentManager.easySend(hostId, cmd);
+        if (answer != null && answer.getResult()) {
+            s_logger.debug("Certificate " + certificateName + "successfully uploaded to host: " + hostId);
```
would be good to have it at the `info`level

marcaurele commented on utils/src/main/java/com/cloud/utils/UriUtils.java:
```diff
@@ -289,6 +297,93 @@ public static long getRemoteSize(String url) {
+        List<String> urls = new ArrayList<>();
+        try {
+            if (httpClient.executeMethod(getMethod) == HttpStatus.SC_OK) {
+                InputStream is = getMethod.getResponseBodyAsStream();
```
same remark for the try with resource statement here. Currently the inputstream close is missing.

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/DirectTemplateDownloaderImpl.java:
```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * Return install full path
+     * @return
```
Empty @Return

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/DirectTemplateDownloaderImpl.java:
```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * Return extract command to execute given downloaded file
+     * @return
```
Empty @Return

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/DirectTemplateDownloaderImpl.java:
```diff
@@ -0,0 +1,193 @@
+     */
+    private String getExtractCommandForDownloadedFile() {
+        if (downloadedFilePath.endsWith(".zip")) {
+            return "unzip -p " + downloadedFilePath + " | cat > " + getInstallFullPath();
```
Are all of these programs required by agent installation?
Otherwise, this will cause a runtime error.

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/DirectTemplateDownloaderImpl.java:
```diff
@@ -0,0 +1,193 @@
+
+    /**
+     * Return checksum command from algorithm
+     * @param algorithm
```
Empty @Return and @Param

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/NfsDirectTemplateDownloader.java:
```diff
@@ -0,0 +1,74 @@
+            setDownloadedFilePath(downloadDir + File.separator + getFileNameFromUrl());
+            Script.runSimpleBashScript("cp /mnt/" + mountSrcUuid + srcPath + " " + getDownloadedFilePath());
+            Script.runSimpleBashScript("umount /mnt/" + mountSrcUuid);
+        } catch (Exception e) {
```
Do you need this generic exception?

nvazquez replied:
Not really, it was just for throwing a checked when download method returns false on KVMStorageProcessor, I'll refactor this

rafaelweingartner commented on engine/orchestration/pom.xml:
```diff
@@ -58,13 +58,13 @@
-    </dependency>
-    -->
+      <!--
+      <dependency>
```
Please remove commented things once you uncover them.

rafaelweingartner commented on engine/storage/image/src/org/apache/cloudstack/storage/image/TemplateDataFactoryImpl.java:
```diff
@@ -153,4 +164,67 @@ public TemplateInfo getReadyTemplateOnCache(long templateId) {
 
+    /**
+     * Given existing spool refs, return one pool id existing on pools and refs
+     * @param existingRefs
```
Remove @Param and @Return that are not properly used

rafaelweingartner commented on engine/storage/image/src/org/apache/cloudstack/storage/image/TemplateDataFactoryImpl.java:
```diff
@@ -153,4 +164,67 @@ public TemplateInfo getReadyTemplateOnCache(long templateId) {
+
+    /**
+     * Retrieve storage pools with scope = cluster or zone matching clusterId or dataCenterId depending on their scope
+     * @param clusterId
```
same here

rafaelweingartner commented on engine/storage/src/org/apache/cloudstack/storage/image/db/TemplateDataStoreDaoImpl.java:
```diff
@@ -549,4 +557,34 @@ public void expireDnldUrlsForZone(Long dcId){
+        sc.setParameters("download_state", Status.BYPASSED);
+        sc.setParameters("state", State.Ready);
+        List<TemplateDataStoreVO> list = search(sc, null);
+        if (list == null || list.isEmpty() || list.size() > 1) {
```
You could use `CollectionUtils.isEmpty() || list.size >1`

rafaelweingartner commented on engine/storage/volume/src/org/apache/cloudstack/storage/volume/VolumeServiceImpl.java:
```diff
@@ -1175,6 +1175,15 @@ private HostVO getHost(Long zoneId, HypervisorType hypervisorType, boolean compu
 
+    @DB
+    @Override
+    public boolean createVolumeFromBypassedTemplate(VolumeInfo volume, long dataStoreId) {
```
Where is this method used?

rafaelweingartner commented on framework/direct-download/src/org/apache/cloudstack/framework/agent/direct/download/DirectDownloadService.java:
```diff
@@ -0,0 +1,36 @@
+
+    /**
+     * Upload client certificate to each running host
+     * @param certificateCer
```
@Param not used

rafaelweingartner commented on framework/direct-download/src/org/apache/cloudstack/framework/agent/direct/download/DirectDownloadService.java:
```diff
@@ -0,0 +1,36 @@
+
+    /**
+     * Download template/ISO into poolId bypassing secondary storage. Download performed by hostId
+     * @param templateId
```
@Param not used

rafaelweingartner commented on plugins/hypervisors/ovm3/src/main/java/com/cloud/hypervisor/ovm3/resources/Ovm3StorageProcessor.java:
```diff
@@ -820,6 +821,12 @@ public ResignatureAnswer resignature(final ResignatureCommand cmd) {
 
+    @Override
+    public Answer handleDownloadTemplateToPrimaryStorage(DirectDownloadCommand cmd) {
+        // Not implemented for OV3
```
These comments can be remove

rafaelweingartner commented on plugins/hypervisors/vmware/src/com/cloud/storage/resource/VmwareStorageProcessor.java:
```diff
@@ -2425,4 +2426,10 @@ public void setFullCloneFlag(boolean value){
+
+    @Override
+    public Answer handleDownloadTemplateToPrimaryStorage(DirectDownloadCommand cmd) {
+        // Not implemented for Vmware
```
comment

rafaelweingartner commented on server/src/com/cloud/resource/ResourceManagerImpl.java:
```diff
@@ -2797,6 +2798,23 @@ public void updateGPUDetails(final long hostId, final HashMap<String, HashMap<St
+        sc.and(sc.entity().getResourceState(), Op.EQ, ResourceState.Enabled);
+        sc.and(sc.entity().getRemoved(), Op.NULL);
+        List<HostVO> hosts = sc.list();
+        if (hosts == null || hosts.isEmpty()) {
```
CollectionUtils.isEmpty

rafaelweingartner commented on server/src/com/cloud/template/HypervisorTemplateAdapter.java:
```diff
@@ -150,6 +185,16 @@ public TemplateProfile prepare(GetUploadParamsForTemplateCmd cmd) throws Resourc
+    /**
+     * Persist template marking it for direct download to Primary Storage, skipping Secondary Storage
+     * @param templateId template id
+     * @param size
```
@param...

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1578

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1580

nvazquez commented:
Thanks a lot @rafaelweingartner and @marcaurele for your valuable reviews

rhtyd commented:
@blueorangutan package
@rafaelweingartner @marcaurele can you re-review the changes/lgtm is okay?

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -551,6 +555,10 @@ protected void processRequest(final Request request, final Link link) {
                         answer = setupAgentKeystore((SetupKeyStoreCommand) cmd);
                     } else if (cmd instanceof SetupCertificateCommand && ((SetupCertificateCommand) cmd).isHandleByAgent()) {
                         answer = setupAgentCertificate((SetupCertificateCommand) cmd);
+                    } else if (cmd instanceof CheckUrlCommand) {
```
This will enable any agent (cpvm, ssvm etc) to handle the command, if we want only the KVM agent to handle these commands they should be moved to libvirt computing resource.

rhtyd commented on agent/src/com/cloud/agent/Agent.java:
```diff
@@ -600,6 +608,47 @@ protected void processRequest(final Request request, final Link link) {
+            return new Answer(cmd, false, "Failed to find agent.properties file");
+        }
+
+        final String keyStoreFile = agentFile.getParent() + "/" + KeyStoreUtils.defaultKeystoreFile;
```
The whole setup process could be implemented as a shell script with the java code calling that script to perform the job.

rafaelweingartner replied:
I agree. You can do that, or you can do everything in Java (reading and writing files).

nvazquez replied:
Yes, I prefer using this java approach for those that can be done instead of using a script file. If you prefer I can refactor it to a script file

rhtyd replied:
I would prefer going the script way, perhaps re-using one of the import keystore/cert scripts. Alternatively, `KeyStore` could be read and manipulated (keys imported etc) by using Java only, agreeing with what @rafaelweingartner suggested.

rhtyd commented on api/src/org/apache/cloudstack/api/command/user/template/RegisterTemplateCmd.java:
```diff
@@ -70,7 +71,7 @@
 
     @Parameter(name = ApiConstants.HYPERVISOR, type = CommandType.STRING, required = true, description = "the target hypervisor for the template")
-    private String hypervisor;
+    protected String hypervisor;
```
This can be retained as private.

rhtyd commented on scripts/vm/hypervisor/kvm/setup_agent.sh:
```diff
@@ -224,5 +224,17 @@ then
+if [ $? -gt 0 ]
+then
+    yum install epel-release -y
+    yum install aria2 -y
```
We can add aria2 dependency on rpms, see the spec files.

nvazquez replied:
I also added it on cloud.spec, should I remove it from here then?

rhtyd commented:
@nvazquez can you also please fix the conflicts?

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-1589

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✖centos6 ✖centos7 ✖debian. JID-1594

borisstoyanov commented:
@nvazquez @DaanHoogland looks like build is still broken, can you have a look please

```
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 53.250s
[INFO] Finished at: Fri Jan 05 10:35:55 UTC 2018
[INFO] Final Memory: 50M/615M
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-checkstyle-plugin:2.17:check (cloudstack-checkstyle) on project cloud-utils: Failed during checkstyle execution: There is 1 error reported by Checkstyle 6.11.2 with cloud-style.xml ruleset. -> [Help 1]
[ERROR] 
```

nvazquez commented:
@borisstoyanov @DaanHoogland I see there are files missing, by splitting the single commit on multiple commits. May I push force with the single commit I'm using to run on my env?

nvazquez commented:
@borisstoyanov @DaanHoogland fixed
@blueorangutan package

blueorangutan commented:
@nvazquez a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1597

rafaelweingartner commented:
@nvazquez I believe you addressed almost everything. From my check there are still the following question in files:
agent/src/com/cloud/agent/direct/download/NfsDirectTemplateDownloader.java 
engine/orchestration/pom.xml

nvazquez commented:
Thanks @rafaelweingartner @rhtyd, I had pushed a refactor commit after your comments

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/DirectTemplateDownloaderImpl.java:
```diff
@@ -0,0 +1,188 @@
+
+    /**
+     * Return direct download temporary path to download template
+     * @param templateId
```
@param and @return can be removed from here

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/DirectTemplateDownloaderImpl.java:
```diff
@@ -0,0 +1,188 @@
+
+    /**
+     * Create folder on path if it does not exist
+     * @param path
```
@param can be removed from here

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/HttpDirectTemplateDownloader.java:
```diff
@@ -0,0 +1,136 @@
+        long offset = 0;
+        boolean done = false;
+        while (!done) {
+            if ((bytes = in.read(block, 0, CHUNK_SIZE)) > -1) {
```
what about using IoUtils to read from one place and write to another?

nvazquez commented:
Thanks @rhtyd @rafaelweingartner, I tried addressing all your comments before going on vacation tomorrow. Thanks @DaanHoogland for taking up the remaining open comment

rafaelweingartner commented on agent/src/com/cloud/agent/direct/download/HttpDirectTemplateDownloader.java:
```diff
@@ -0,0 +1,129 @@
+
+    protected void copyBytes(InputStream in, OutputStream out) throws IOException {
+        byte[] block = new byte[CHUNK_SIZE];
+        while (IOUtils.read(in, block) > 0) {
```
You do not need this while. As a matter of fact, you do not need this method. You can simply use `IOUtils.read(in, out)` at line 115

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1600

DaanHoogland commented:
adressed the final remark, awaiting ci and if our esteemed RM agrees we can merge.
/cc @borisstoyanov @rafaelweingartner @rhtyd @marcaurele ?

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1602

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

rhtyd commented:
Thanks @DaanHoogland I'll check again tomorrow when the results are back, may tackle outstanding issues myself.

blueorangutan commented:
<b>Trillian test result (tid-2046)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 38525 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2379-t2046-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_deploy_virtio_scsi_vm.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Smoke tests completed. 64 look OK, 3 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_register_template_direct_download_flag | `Error` | 0.04 | test_templates.py
test_02_deploy_vm_from_direct_download_template | `Error` | 0.00 | test_templates.py
test_03_deploy_vm_wrong_checksum | `Error` | 0.03 | test_templates.py
test_07_resize_fail | `Failure` | 15.38 | test_volumes.py
test_04_rvpc_network_garbage_collector_nics | `Failure` | 285.37 | test_vpc_redundant.py

rhtyd commented:
@blueorangutan package

blueorangutan commented:
@rhtyd a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1610

rhtyd commented:
@blueorangutan test

blueorangutan commented:
@rhtyd a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2063)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 44857 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2379-t2063-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_public_ip_range.py
Intermitten failure detected: /marvin/tests/smoke/test_ssvm.py
Intermitten failure detected: /marvin/tests/smoke/test_templates.py
Intermitten failure detected: /marvin/tests/smoke/test_usage.py
Intermitten failure detected: /marvin/tests/smoke/test_volumes.py
Intermitten failure detected: /marvin/tests/smoke/test_hostha_kvm.py
Smoke tests completed. 63 look OK, 4 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_01_list_sec_storage_vm | `Failure` | 0.13 | test_ssvm.py
test_02_list_cpvm_vm | `Failure` | 0.14 | test_ssvm.py
test_05_stop_ssvm | `Failure` | 99.28 | test_ssvm.py
test_06_stop_cpvm | `Failure` | 112.30 | test_ssvm.py
test_01_register_template_direct_download_flag | `Error` | 0.04 | test_templates.py
test_02_deploy_vm_from_direct_download_template | `Error` | 0.00 | test_templates.py
test_03_deploy_vm_wrong_checksum | `Error` | 0.04 | test_templates.py
test_04_extract_template | `Failure` | 132.38 | test_templates.py
ContextSuite context=TestISOUsage>:setup | `Error` | 0.00 | test_usage.py
test_06_download_detached_volume | `Failure` | 147.53 | test_volumes.py
test_07_resize_fail | `Failure` | 15.42 | test_volumes.py

borisstoyanov commented:
SSVM tests are not related to this, let me have a quick look at the template ones.

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1620

borisstoyanov commented:
@blueorangutan package

blueorangutan commented:
@borisstoyanov a Jenkins job has been kicked to build packages. I'll keep you posted as I make progress.

blueorangutan commented:
Packaging result: ✔centos6 ✔centos7 ✔debian. JID-1623

borisstoyanov commented:
@blueorangutan test

blueorangutan commented:
@borisstoyanov a Trillian-Jenkins test job (centos7 mgmt + kvm-centos7) has been kicked to run smoke tests

blueorangutan commented:
<b>Trillian test result (tid-2085)</b>
Environment: kvm-centos7 (x2), Advanced Networking with Mgmt server 7
Total time taken: 32518 seconds
Marvin logs: https://github.com/blueorangutan/acs-prs/releases/download/trillian/pr2379-t2085-kvm-centos7.zip
Intermitten failure detected: /marvin/tests/smoke/test_internal_lb.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_redundant.py
Intermitten failure detected: /marvin/tests/smoke/test_vpc_vpn.py
Smoke tests completed. 66 look OK, 1 have error(s)
Only failed tests results shown below:


Test | Result | Time (s) | Test File
--- | --- | --- | ---
test_05_rvpc_multi_tiers | `Failure` | 345.94 | test_vpc_redundant.py
test_05_rvpc_multi_tiers | `Error` | 391.55 | test_vpc_redundant.py

rhtyd commented:
Tests LGTM (template related errors not seen in last run), merging this based on code reviews and test results.
